sklearn/feature_selection/mutual_info.py

```diff
@@ -0,0 +1,375 @@
+        Ranks of features, values from 0 (best) to n_features-1 (worst).
+        If ``use_redundancy=True``, then only `n_features_to_select` values
+        were actually filled, the rest are filled with n_features - 1.
+
+    References
+    ----------
+    .. [1] H. Peng, F. Long, and C. Ding, "Feature selection based on mutual
+           information: criteria of max-dependency, max-relevance, and
+           min-redundancy", IEEE Transactions on Pattern Analysis and Machine
+           Intelligence, Vol. 27, No. 8, pp. 1226-1238, 2005.
+    .. [2] A. Kraskov, H. Stogbauer and P. Grassberger, "Estimating mutual
+           information". Phys. Rev. E 69, 2004.
+    .. [3] B. C. Ross "Mutual Information between Discrete and Continuous
+           Data Sets". PLoS ONE 9(2), 2014.
+    """
+    def __init__(self, n_features_to_select=0.5, use_redundancy=True,
+                 categorical_features=False, categorical_target=False,
+                 n_neighbors=3, copy=True):
+        self.n_features_to_select = n_features_to_select
+        self.use_redundancy = use_redundancy
+        self.categorical_features = categorical_features
+        self.categorical_target = categorical_target
+        self.n_neighbors = n_neighbors
+        self.copy = copy
+
+    def _get_support_mask(self):
+        n_features_to_select = self._get_n_features_to_select()
+        support = np.zeros_like(self.ranks_, dtype=bool)
+        support[self.ranks_ < n_features_to_select] = True
+        return support
+
+    def _get_n_features_to_select(self):
+        if self.n_features_to_select >= 1:
+            return int(self.n_features_to_select)
+        elif 0 < self.n_features_to_select < 1:
+            return max(1, int(self.n_features_to_select * self._n_features))
+        else:
+            raise ValueError("`n_features_to_select` must be positive.")
+
+    def fit(self, X, y):
+        """Find the best features according to mutual information criterion.
+
+        Parameters
+        ----------
+        X : array_like or sparse matrix, shape (n_samples, n_features)
+            Feature matrix.
+        y : array_like, shape (n_samples,)
+            Target vector.
+
+        Returns
+        -------
+        self
+        """
+        X, y = check_X_y(X, y, accept_sparse='csc',
+                         y_numeric=not self.categorical_target)
+
+        n_features = X.shape[1]
+        self._n_features = n_features
+        n_features_to_select = self._get_n_features_to_select()
+
+        if isinstance(self.categorical_features, bool):
+            categorical_features = np.empty(n_features, dtype=bool)
+            categorical_features.fill(self.categorical_features)
+        else:
+            categorical_features = np.asarray(self.categorical_features)
+
+        if (not np.all(categorical_features) and
+                (self.use_redundancy or not self.categorical_target)):
+            X = scale(X, with_mean=False, copy=self.copy)
```

Comment:
If that is your concern, then we are creating quite a few copies of the redundancy array below anyway by doing non-contiguous slices.